src/features/DocumentFormatter.ts

```diff
@@ -0,0 +1,277 @@
+
+export namespace ScriptFileMarkersRequest {
+    export const type: RequestType<any, any, void> = { get method(): string { return "powerShell/getScriptFileMarkers"; } };
+}
+
+// TODO move some of the common interface to a separate file?
+interface ScriptFileMarkersRequestParams {
+    filePath: string;
+    settings: string;
+}
+
+interface ScriptFileMarkersRequestResultParams {
+    markers: ScriptFileMarker[];
+}
+
+interface ScriptFileMarker {
+    message: string;
+    level: ScriptFileMarkerLevel;
+    scriptRegion: ScriptRegion;
+    correction: MarkerCorrection;
+}
+
+enum ScriptFileMarkerLevel {
+    Information = 0,
+    Warning,
+    Error
+}
+
+interface ScriptRegion {
+    file: string;
+    text: string;
+    startLineNumber: number;
+    startColumnNumber: number;
+    startOffset: number;
+    endLineNumber: number;
+    endColumnNumber: number;
+    endOffset: number;
+}
+
+interface MarkerCorrection {
+    name: string;
+    edits: ScriptRegion[]
+}
+
+function editComparer(leftOperand: ScriptRegion, rightOperand: ScriptRegion): number {
+    if (leftOperand.startLineNumber < rightOperand.startLineNumber) {
+        return -1;
+    } else if (leftOperand.startLineNumber > rightOperand.startLineNumber) {
+        return 1;
+    } else {
+        if (leftOperand.startColumnNumber < rightOperand.startColumnNumber) {
+            return -1;
+        }
+        else if (leftOperand.startColumnNumber > rightOperand.startColumnNumber) {
+            return 1;
+        }
+        else {
+            return 0;
+        }
+    }
+}
+
+class PSDocumentFormattingEditProvider implements DocumentFormattingEditProvider, DocumentRangeFormattingEditProvider {
+    private languageClient: LanguageClient;
+
+    // The order in which the rules will be executed starting from the first element.
+    private readonly ruleOrder: string[] = [
+        "PSPlaceCloseBrace",
+        "PSPlaceOpenBrace",
+        "PSUseConsistentIndentation"];
+
+    // Allows edits to be undone and redone is a single step.
+    // Should we expose this through settings?
+    private aggregateUndoStop: boolean;
+
+    constructor(aggregateUndoStop: boolean) {
+        this.aggregateUndoStop = aggregateUndoStop;
+    }
+
+    provideDocumentFormattingEdits(
+        document: TextDocument,
+        options: FormattingOptions,
+        token: CancellationToken): TextEdit[] | Thenable<TextEdit[]> {
+        return this.provideDocumentRangeFormattingEdits(document, null, options, token);
+    }
+
+    provideDocumentRangeFormattingEdits(
+        document: TextDocument,
+        range: Range,
+        options: FormattingOptions,
+        token: CancellationToken): TextEdit[] | Thenable<TextEdit[]> {
+        return this.executeRulesInOrder(document, range, options, 0);
+    }
+
+    executeRulesInOrder(
+        document: TextDocument,
+        range: Range,
+        options: FormattingOptions,
+        index: number): Thenable<TextEdit[]> | TextEdit[] {
```

Comment:
I think that we could simplify this code a lot by using the TextEdit class as the means for transporting the formatting edits back and forth between extension and server.  Check out VS Code's [TextEdit](https://code.visualstudio.com/docs/extensionAPI/vscode-api#_a-nametexteditaspan-classcodeitem-id586texteditspan) class compared to the one we have in the [Protocol assembly](https://github.com/PowerShell/PowerShellEditorServices/blob/develop/src/PowerShellEditorServices.Protocol/LanguageServer/Completion.cs#L48), they are exactly the same.

I'd recommend pushing as much of this logic to the service side as you can and only sending back the valid TextEdits from the server which will be applied by the client.  In fact, I think we should implement this based on the exact protocol messages that are coming in the Language Server Protocol 3.0 since we'll be snapping to that pretty soon anyway.  Check these out:

https://github.com/Microsoft/language-server-protocol/blob/master/versions/protocol-2-x.md#textDocument_formatting

The request type is fairly similar to yours now and the response type is simply an array of `TextEdit`.  Implementing it this way will future-proof us a lot better.